cpp/src/quantiles/quantiles.cu

```diff
@@ -16,196 +16,253 @@
-          {
-            RetT rt = quantile_exact(d_in.data().get(), d_in.size(), q, vf[i]);
-            std::cout<<"q: "<<q<<"; method: "<<methods[i]<<"; rt: "<<rt<<"\n";
-          }
-      }
+    if( quant >= 1.0 && !flag_sorted )
+    {
+      T* d_res = thrust::max_element(rmm::exec_policy(stream)->on(stream), dv, dv+n);
+      singleMemcpy(hv[0], d_res, stream);
+      result = static_cast<RetT>( hv[0] );
+      return GDF_SUCCESS;
+    }
+
+    if( quant <= 0.0 && !flag_sorted )
+    {
+      T* d_res = thrust::min_element(rmm::exec_policy(stream)->on(stream), dv, dv+n);
+      singleMemcpy(hv[0], d_res, stream);
+      result = static_cast<RetT>( hv[0] );
+      return GDF_SUCCESS;
+    }
+
+    // sort if the input is not sorted.
+    if( !flag_sorted ){
+      thrust::sort(rmm::exec_policy(stream)->on(stream), dv, dv+n);
+    }
+
+    QuantiledIndex qi = find_quantile_index(n, quant);
+
+    switch( interpolation )
+    {
+    case GDF_QUANT_LINEAR:
+      singleMemcpy(hv[0], dv+qi.lower_bound, stream);
+      singleMemcpy(hv[1], dv+qi.upper_bound, stream);
+      cudf::interpolate::linear(
+          cudf::detail::unwrap(result), 
+          cudf::detail::unwrap(hv[0]), 
+          cudf::detail::unwrap(hv[1]), qi.fraction);
+      break;
+    case GDF_QUANT_MIDPOINT:
+      singleMemcpy(hv[0], dv+qi.lower_bound, stream);
+      singleMemcpy(hv[1], dv+qi.upper_bound, stream);
+      cudf::interpolate::midpoint(
+          cudf::detail::unwrap(result), 
+          cudf::detail::unwrap(hv[0]), 
+          cudf::detail::unwrap(hv[1]));
+      break;
+    case GDF_QUANT_LOWER:
+      singleMemcpy(hv[0], dv+qi.lower_bound, stream);
+      result = static_cast<RetT>( hv[0] );
+      break;
+    case GDF_QUANT_HIGHER:
+      singleMemcpy(hv[0], dv+qi.upper_bound, stream);
+      result = static_cast<RetT>( hv[0] );
+      break;
+    case GDF_QUANT_NEAREST:
+      singleMemcpy(hv[0], dv+qi.nearest, stream);
+      result = static_cast<RetT>( hv[0] );
+      break;
+
+    default:
+      return GDF_UNSUPPORTED_METHOD;
+    }
+    
+    return GDF_SUCCESS;
   }
 
   template<typename ColType,
-           typename RetT = double> // just in case double won't be enough to hold result, in the future
+           typename RetT = double>
   gdf_error trampoline_exact(gdf_column*  col_in,
-                             gdf_quantile_method prec,
-                             double q,
+                             gdf_quantile_method interpolation,
+                             double quant,
                              void* t_erased_res,
-                             gdf_context* ctxt)
+                             gdf_context* ctxt,
+                             cudaStream_t stream = NULL)
   {
     RetT* ptr_res = static_cast<RetT*>(t_erased_res);
     size_t n = col_in->size;
     ColType* p_dv = static_cast<ColType*>(col_in->data);
+    
     if( ctxt->flag_sort_inplace || ctxt->flag_sorted)
       {
         return select_quantile(p_dv,

```

Comment:
Seems the call to select_quantile is identical in the if and else. So perhaps just move it after the if/else and just set the options up in the if/else.